datacube_stats/utils/__init__.py

```diff
@@ -96,6 +96,53 @@ def to_float(da):
     return data.apply(to_float, keep_attrs=True)
 
 
+def cast_back(data, measurements):
+    """
+    Cast calculated statistic `Dataset` into intended data types.
+    When going through intermediate representation as floats,
+    restore `nodata` values in place of `NaN`s.
+    """
+    assert isinstance(data, xarray.Dataset)
+    measurements = {measurement['name']: measurement
+                    for measurement in measurements}
+
+    data_vars = [name for name in data.data_vars]
+    assert set(data_vars) == set(measurements.keys())
+
+    def cast(da):
+        """ Cast `DataArray` into intended type. """
+        output_measurement = measurements[da.name]
+        expected_dtype = np.dtype(output_measurement['dtype'])
+        actual_dtype = da.dtype
+
+        # our intermediate data type
+        float32 = np.dtype(np.float32)
+
+        if actual_dtype != float32:
```

Comment:
I think this assumes too much about the context in which this code is used, why shouldn't I be able to cast float64? We support custom external plugins, users might choose to use float64 in some situations, or by mistake.